src/main/java/org/jruby/ext/openssl/OCSPRequest.java

```diff
@@ -0,0 +1,169 @@
+    private List<OCSPCertificateId> certificateIds;
+    private byte[] nonce;
+    private SecureRandom random = new SecureRandom();
+    
+    @JRubyMethod(name = "initialize", rest = true, visibility = Visibility.PRIVATE)
+    public IRubyObject initialize(final ThreadContext context, IRubyObject[] args) {
+        Ruby runtime = context.getRuntime();
+        
+        if ( Arity.checkArgumentCount(runtime, args, 0, 1) == 0 ) return this;
+        
+        RubyString derString = (RubyString) args[0];
+        try {
+            bcReq = org.bouncycastle.asn1.ocsp.OCSPRequest.getInstance(ASN1TaggedObject.fromByteArray(derString.getBytes()));
+        }
+        catch (IOException e) {
+            throw newOCSPError(runtime, e);
+        }
+        
+        return this;
+    }
+    
+    @JRubyMethod(name = "add_certid")
+    public IRubyObject add_certid(IRubyObject certId) {
+        OCSPCertificateId rubyCertId = (OCSPCertificateId) certId;
+        certificateIds.add(rubyCertId);
+        
+        return this;
+    }
+    
+    @JRubyMethod(name = "add_nonce")
+    public IRubyObject add_nonce(IRubyObject[] args) {
+        Ruby runtime = getRuntime();
+        
+        if ( Arity.checkArgumentCount(runtime, args, 0, 1) == 0 ) {
+            nonce = generateNonce();
+        }
+        else {
+            RubyString input = (RubyString)args[0];
+            nonce = generateNonce(input.getBytes());
+        }
+        return this;
+    }
+    
+    @JRubyMethod(name = "certid")
+    public IRubyObject certid() {
+        return RubyArray.newArray(getRuntime(), certificateIds);
+    }
+    
+    @JRubyMethod(name = "check_nonce")
+    public IRubyObject check_nonce(IRubyObject response) {
+        Ruby runtime = getRuntime();
+        if (response instanceof OCSPBasicResponse) {
+            OCSPBasicResponse rubyBasicRes = (OCSPBasicResponse) response;
+            return checkNonceImpl(runtime, this.nonce, rubyBasicRes.getNonce());
+        }
+        else if (response instanceof OCSPResponse) {
+            OCSPResponse rubyResp = (OCSPResponse) response;
+            return checkNonceImpl(runtime, this.nonce, rubyResp.getBasicResponse().getNonce());
+        }
+        else {
+            return checkNonceImpl(runtime, this.nonce, null);
+        }
+    }
+    
+    @JRubyMethod(name = "sign", rest = true)
+    public IRubyObject sign(IRubyObject cert, IRubyObject key, IRubyObject[] args) {
+        //TODO finish
+        X509Cert rubyCert = (X509Cert) cert;
+        
+        return this;
+    }
+    
+    @JRubyMethod(name = "to_der")
+    public IRubyObject to_der() {
+        //TODO implement
+        return RubyString.newEmptyString(getRuntime());
+    }
+    
+    @JRubyMethod(name = "verify")
+    public IRubyObject verify(IRubyObject certificates, IRubyObject store, IRubyObject flags) {
+        //TODO implement
+        
+        return RubyBoolean.newBoolean(getRuntime(), true);
```

Comment:
safer to return `false` if its unimplemented for users relying on this feature, right?